src/foolscap/storage.py

```diff
@@ -294,14 +354,22 @@ def receiveChild(self, obj, ready_deferred=None):
         assert not isinstance(obj, Deferred)
         assert ready_deferred is None
         if self.state == 0:
-            self.im_func = obj
+            self.im_func = six.ensure_str(obj)
             self.state = 1
         elif self.state == 1:
-            assert type(obj) in (types.InstanceType, types.NoneType)
+            if six.PY2:
+              assert type(obj) in (InstanceType, type(None))
+            elif six.PY3:
+              # No more instance types in PY3
+              if hasattr(obj, '__class__'):
+                assert(isinstance(obj.__class__, type))
+              else:
+                assert(type(obj) == type(None))
+                
```

Comment:
if these are critical checks, we shouldn't use `assert` because it isn't run in all cases. e.g. use `if not hasattr(): raise ..` instead.